linera-core/src/updater.rs

```diff
@@ -179,80 +183,75 @@ where
     sample.truncate(4);
-    Err(CommunicationError::Sample(sample))
+    Err(CommunicationError::Sample(sample).into())
 }
 
 impl<A, S> ValidatorUpdater<A, S>
 where
     A: ValidatorNode + Clone + 'static,
     S: Storage + Clone + Send + Sync + 'static,
 {
-    async fn send_optimized_certificate(
-        &mut self,
-        certificate: &Certificate,
-        delivery: CrossChainMessageDelivery,
-    ) -> Result<Box<ChainInfo>, NodeError> {
-        if certificate.is_signed_by(&self.remote_node.name) {
-            let result = self
-                .remote_node
-                .handle_lite_certificate(certificate.lite_certificate(), delivery)
-                .await;
-            match result {
-                Err(NodeError::MissingCertificateValue) => {
-                    warn!(
-                        "Validator {} forgot a certificate value that they signed before",
-                        self.remote_node.name
-                    );
-                }
-                _ => {
-                    return result;
-                }
-            }
-        }
-        self.remote_node
-            .handle_certificate(certificate.clone(), vec![], delivery)
-            .await
-    }
-
     async fn send_certificate(
         &mut self,
         certificate: Certificate,
         delivery: CrossChainMessageDelivery,
-    ) -> Result<Box<ChainInfo>, NodeError> {
+    ) -> Result<Box<ChainInfo>, ChainClientError> {
         let result = self
-            .send_optimized_certificate(&certificate, delivery)
+            .remote_node
+            .handle_optimized_certificate(&certificate, delivery)
             .await;
 
         match &result {
             Err(original_err @ NodeError::BlobsNotFound(blob_ids)) => {
+                // Find the missing blobs locally and retry.
+                let required = match certificate.value() {
+                    CertificateValue::ConfirmedBlock { executed_block, .. }
+                    | CertificateValue::ValidatedBlock { executed_block, .. } => {
+                        executed_block.required_blob_ids()
+                    }
+                    CertificateValue::Timeout { .. } => HashSet::new(),
+                };
+                for blob_id in blob_ids {
+                    if !required.contains(blob_id) {
+                        warn!(
+                            "validator requested blob {:?} but it is not required",
+                            blob_id
+                        );
+                        return Err(NodeError::UnexpectedEntriesInBlobsNotFound.into());
+                    }
+                }
+                let unique_missing_blob_ids = blob_ids.iter().cloned().collect::<HashSet<_>>();
+                if blob_ids.len() > unique_missing_blob_ids.len() {
+                    warn!("blobs requested by validator contain duplicates");
+                    return Err(NodeError::DuplicatesInBlobsNotFound.into());
+                }
```

Comment:
This part isn't mentioned in the PR description ;)

Reply:
Right, this was moved out of `LocalNode::find_missing_blobs`, because these ones _are_ the fault of the remote node.